quantum/process_keycode/process_auto_shift.c

```diff
@@ -33,42 +46,178 @@ void autoshift_timer_report(void) {
+        // Allow a tap-then-hold to hold the unshifted key.
+        autoshift_lastkey = KC_NO;
+        return true;
+    }
+#    endif
+
+    // Record the keycode so we can simulate it later.
+    autoshift_lastkey           = keycode;
+    autoshift_flags.in_progress = true;
+
+#    if !defined(NO_ACTION_ONESHOT) && !defined(NO_ACTION_TAPPING)
+    clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
+#    endif
+    return false;
+}
+
+/** \brief Registers an autoshiftable key under the right conditions
+ *
+ * If the autoshift delay has elapsed and no shift has already been registered,
+ * register a shift and the key.
+ *
+ * \return True if the timeout had elapsed.
+ */
+void autoshift_check_timeout(uint16_t now) {
+    const uint16_t elapsed = TIMER_DIFF_16(now, autoshift_time);
+    if (elapsed >= autoshift_timeout) {
+        // The timeout has expired, so simulate the keypress.
+        if (!(get_mods() & (MOD_BIT(KC_LSFT) | MOD_BIT(KC_RSFT)))) {
+            // Simulate pressing the shift key.
+            register_code(KC_LSFT);
+            autoshift_flags.lastshifted   = true;
+            autoshift_flags.holding_shift = true;
+        }
+        register_code(autoshift_lastkey);
+        autoshift_flags.registered = true;
+
+#    if TAP_CODE_DELAY > 0
+        wait_ms(TAP_CODE_DELAY);
+#    endif
+    }
 }
 
-void autoshift_flush(void) {
-    if (autoshift_lastkey != KC_NO) {
-        uint16_t elapsed = timer_elapsed(autoshift_time);
+/** \brief Registers an autoshiftable key under the right conditions
+ *
+ * If the autoshift delay has elapsed and no shift has already been registered,
+ * register a shift and the key.
+ *
+ * If the autoshift key is released before the delay has elapsed, register the
+ * key without a shift.
+ *
+ * If another key is pressed, register the key.
+ */
+static void autoshift_check_record(uint16_t keycode, keyrecord_t *record) {
+    if (!autoshift_flags.in_progress) {
+        return;
+    }
 
-        if (elapsed > autoshift_timeout) {
-            tap_code16(LSFT(autoshift_lastkey));
+    // Process the release of the autoshiftable key.
+    if (keycode == autoshift_lastkey && !record->event.pressed) {
+        autoshift_flush_shift();
+        // Time since the initial press was recorded.
+        const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
+        if (!autoshift_flags.registered && elapsed < autoshift_timeout) {
+            // Auto-shiftable key is being released before the shift timeout;
+            // simulate the original press then let the usual processing take
+            // care of the release.
+            register_code(keycode);
+            autoshift_flags.lastshifted = false;
+            autoshift_flags.registered  = true;
+#    if TAP_CODE_DELAY > 0
+            wait_ms(TAP_CODE_DELAY);
+#    endif
+        }
+        autoshift_flush(record->event.time);
+    } else if (record->event.pressed) {
+        if (keycode == KC_LSFT) {
```

Comment:
This needs right shift added if my cleanup isn't merged.
```suggestion
        if (keycode == KC_LSFT || keycode == KC_RSFT) {
```

Reply:
The auto-shifting here only ever holds down LSFT.  If the user presses RSFT, it shouldn't interfere with auto-shift and can just be processed normally.

Reply:
I thought that holding LSFT, holding a key, and releasing would unshift while RSFT would not. I don't know why, but I just tested and that's not the case. Tapping either while repeating shifted also unshifts. You're right, not an issue.